scipy/optimize/_qap.py

```diff
@@ -0,0 +1,418 @@
+           matching," PLOS one, vol. 10, no. 4, p. e0121002, 2015.
+
+    .. [2] D. Fishkind, S. Adali, H. Patsolic, L. Meng, D. Singh, V. Lyzinski,
+           C. Priebe, "Seeded graph matching", Pattern Recognit. 87 (2019):
+           203-215.
+
+    Examples
+    --------
+
+    >>> cost = np.array([[0, 80, 150, 170], [80, 0, 130, 100],
+    ...         [150, 130, 0, 120], [170, 100, 120, 0]])
+    >>> dist = np.array([[0, 5, 2, 7], [0, 0, 3, 8],
+    ...         [0, 0, 0, 3], [0, 0, 0, 0]])
+    >>> from scipy.optimize import quadratic_assignment
+    >>> res = quadratic_assignment(cost, dist)
+    >>> print(res)
+     col_ind: array([0, 3, 2, 1])
+         nit: 9
+       score: 3260
+
+    To demonstrate explicitly how the `score` value
+    :math:`f(P) = trace(A^T PBP^T )` is calculated, one may construct the
+    permutation matrix, and perform the necessary algebra.
+
+    >>> n = cost.shape[0]
+    >>> P = np.zeros((n, n))
+    >>> P[np.arange(n), res['col_ind']] = 1
+    >>> score = int(np.trace(cost.T @ P @ dist @ P.T))
+    >>> print(score)
+    3260
+
+    As you can see, the value here matches res['score'] reported above.
+    Alternatively, to avoid constructing the permutation matrix, one can also
+    perform the following calculation.
+
+    >>> score = np.trace(cost.T @ dist[np.ix_(res['col_ind'], res['col_ind'])])
+    >>> print(score)
+    3260
+
+    Here, we are simply permuting the distance matrix.
+
+    """
+
+    cost_matrix = np.asarray(cost_matrix)
+    dist_matrix = np.asarray(dist_matrix)
+
+    if seed is None:
+        seed = np.array([[], []]).T
+    seed = np.asarray(seed)
+
+    # ValueError check
+    msg = None
+    if cost_matrix.shape[0] != dist_matrix.shape[0]:
+        msg = "Adjacency matrices must be of equal size"
+    elif (
+        cost_matrix.shape[0] != cost_matrix.shape[1]
+        or dist_matrix.shape[0] != dist_matrix.shape[1]
+    ):
+        msg = "Adjacency matrix entries must be square"
+    elif not (cost_matrix >= 0).all() or not (dist_matrix >= 0).all():
+        msg = "Adjacency matrix entries must be greater than or equal to zero"
+    elif seed.shape[0] > cost_matrix.shape[0]:
+        msg = "There cannot be more seeds than there are nodes"
+    elif seed.shape[1] != 2:
+        msg = "Seed array entry must have two columns"
+    elif not (seed >= 0).all():
+        msg = "Seed array entries must be greater than or equal to zero"
+    elif not (seed <= (cost_matrix.shape[0] - 1)).all():
+        msg = "Seed array entries must be less than the number of nodes"
+    elif not len(set(seed[:, 0])) == len(seed[:, 0]) or not \
+            len(set(seed[:, 1])) == len(seed[:, 1]):
+        msg = "Seed column entries must be unique"
+    elif init_method not in {'barycenter', 'rand'}:
+        msg = "Invalid 'init_method' parameter string"
+    if msg is not None:
+        raise ValueError(msg)
+
+    # TypeError check
+    if type(n_init) is not int or n_init <= 0:
```

Comment:
The checks for `int` and `float` types are too strict. A user could provide a `numpy.int64` type, which is valid. I suggest you cast the input variables into your preferred type, i.e.:
```
n_init = int(n_init)
```
The remaining check `n_init <= 0` is should then raise a `ValueError` rather than `TypeError`.

Reply:
FYI: These days the method that we've been using to check that an input is, in fact, an integer is to pass the parameter through `operator.index`.  That will accept all the numpy integer types, but will raise a `TypeError` with, for example, a floating point value.